crates/viewer/re_renderer/src/renderer/video.rs

```diff
@@ -0,0 +1,544 @@
+    ///
+    /// If the timestamp is negative, a zeroed texture is returned.
+    pub fn get_frame(&mut self, timestamp_s: f64) -> GpuTexture2D {
+        re_tracing::profile_function!();
+        self.decoder.get_frame(TimeMs::new(timestamp_s * 1e3))
+    }
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum VideoError {
+    #[error("{0}")]
+    Load(#[from] VideoLoadError),
+
+    #[error("failed to initialize video decoder")]
+    Init,
+}
+
+use decoder::VideoDecoder;
+
+fn alloc_video_frame_texture(
+    device: &wgpu::Device,
+    pool: &GpuTexturePool,
+    width: u32,
+    height: u32,
+) -> GpuTexture2D {
+    let Some(texture) = GpuTexture2D::new(pool.alloc(
+        device,
+        &TextureDesc {
+            label: "video".into(),
+            size: wgpu::Extent3d {
+                width,
+                height,
+                depth_or_array_layers: 1,
+            },
+            mip_level_count: 1,
+            sample_count: 1,
+            dimension: wgpu::TextureDimension::D2,
+            format: wgpu::TextureFormat::Rgba8Unorm,
+            usage: wgpu::TextureUsages::COPY_DST
+                | wgpu::TextureUsages::TEXTURE_BINDING
+                | wgpu::TextureUsages::RENDER_ATTACHMENT,
+        },
+    )) else {
+        unreachable!();
+    };
+
+    texture
+}
+
+/// Returns the index of:
+/// - The index of `needle` in `v`, if it exists
+/// - The index of the first element in `v` that is lesser than `needle`, if it exists
+/// - `None`, if `v` is empty OR `needle` is greater than all elements in `v`
+fn latest_at_idx<T, K: Ord>(v: &[T], key: impl Fn(&T) -> K, needle: &K) -> Option<usize> {
+    if v.is_empty() {
+        return None;
+    }
+
+    let idx = v.partition_point(|x| key(x) <= *needle);
+
+    if idx == 0 {
+        // If idx is 0, then all elements are greater than the needle
+        if &key(&v[0]) > needle {
+            return None;
+        }
+    }
+
+    Some(idx.saturating_sub(1))
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_latest_at_idx() {
+        let v = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
+        assert_eq!(latest_at_idx(&v, |v| *v, &0), None);
+        assert_eq!(latest_at_idx(&v, |v| *v, &1), Some(0));
+        assert_eq!(latest_at_idx(&v, |v| *v, &2), Some(1));
+        assert_eq!(latest_at_idx(&v, |v| *v, &3), Some(2));
+        assert_eq!(latest_at_idx(&v, |v| *v, &4), Some(3));
+        assert_eq!(latest_at_idx(&v, |v| *v, &5), Some(4));
+        assert_eq!(latest_at_idx(&v, |v| *v, &6), Some(5));
+        assert_eq!(latest_at_idx(&v, |v| *v, &7), Some(6));
+        assert_eq!(latest_at_idx(&v, |v| *v, &8), Some(7));
+        assert_eq!(latest_at_idx(&v, |v| *v, &9), Some(8));
+        assert_eq!(latest_at_idx(&v, |v| *v, &10), Some(9));
+        assert_eq!(latest_at_idx(&v, |v| *v, &11), Some(9));
+        assert_eq!(latest_at_idx(&v, |v| *v, &1000), Some(9));
+    }
+}
+
+#[cfg(target_arch = "wasm32")]
+mod decoder {
```

Comment:
This could move to its own file `video/web.rs`, and make `video.rs` a `video/mod.rs`. We're only expecting this module to grow, especially once we add native support.